pkg/traceability/muleemitter.go

```diff
@@ -2,89 +2,150 @@ package traceability
+		return err
+	}
 
-// pollForEvents - Polls for the events
-func (me *MuleEventEmitter) pollForEvents() {
-	ticker := time.NewTicker(me.pollInterval)
-	go func() {
-		for {
-			select {
-			case <-me.done:
-				return
-			case <-ticker.C:
-				oldTime := time.Now()
-				events, err := me.client.GetAnalyticsWindow()
-				currentTime := time.Now()
-				duration := currentTime.Sub(oldTime)
-				logrus.WithFields(logrus.Fields{
-					"duration": duration * time.Millisecond,
-					"count":    len(events),
-				}).Debug("retrieved events from anypoint")
-				if err != nil {
-					logp.Warn("Client Failure: %s", err.Error())
-				}
-				for _, event := range events {
-					j, err := json.Marshal(event)
-					if err != nil {
-						logp.Warn("Marshal Failure: %s", err.Error())
-					}
-					me.eventChannel <- string(j)
-				}
-			}
+	for _, event := range events {
+		j, err := json.Marshal(event)
+		if err != nil {
+			log.Warnf("failed to marshal event: %s", err.Error())
 		}
-	}()
-}
-
-// Stop -
-func (me *MuleEventEmitter) Stop() {
-	me.done <- true
+		me.eventChannel <- string(j)
+	}
+	return nil
 }
 
 // OnConfigChange -
 func (me *MuleEventEmitter) OnConfigChange(gatewayCfg *config.AgentConfig) {
+	me.pollInterval = gatewayCfg.MulesoftConfig.PollInterval
 	me.client.OnConfigChange(gatewayCfg.MulesoftConfig)
 }
+
+// NewMuleEventEmitterJob creates a struct that implements the Emitter and Job interfaces.
+func NewMuleEventEmitterJob(
+	emitter Emitter,
+	pollInterval time.Duration,
+	healthCheck hc.CheckStatus,
+) (*MuleEventEmitterJob, error) {
+	_, err := hc.RegisterHealthcheck("Data Ingestion Endpoint", healthCheckEndpoint, healthCheck)
+	if err != nil {
+		return nil, err
+	}
+	return &MuleEventEmitterJob{
+		Emitter:      emitter,
+		pollInterval: pollInterval,
+	}, nil
+}
+
+// Start registers the job with the sdk.
+func (m *MuleEventEmitterJob) Start() error {
+	jobID, err := jobs.RegisterIntervalJob(m, m.pollInterval)
+	m.jobID = jobID
+	return err
+}
+
+// Execute called by the sdk on each interval.
+func (m *MuleEventEmitterJob) Execute() error {
+	return m.Emitter.Start()
+}
+
+// Status Performs a health check for this job before it is executed.
+func (m *MuleEventEmitterJob) Status() error {
+	max := 3
+	status := hc.GetStatus(healthCheckEndpoint)
+	if status == hc.OK {
+		m.consecutiveErrors = 0
+		return nil
+	} else if m.consecutiveErrors >= max {
+		// If the job fails 3 times return an error
+		return fmt.Errorf("failed to start the Traceability agent %d times in a row", max)
+	}
+	m.consecutiveErrors++
+	return nil
+}
+
+// Ready determines if the job is ready to run.
+func (m *MuleEventEmitterJob) Ready() bool {
+	status := hc.GetStatus(healthCheckEndpoint)
+	if status == hc.OK {
+		return true
+	}
+	return false
+}
+
+func traceabilityHealthCheck(name string) *hc.Status {
+	status := &hc.Status{
+		Result: hc.OK,
+	}
+	// Check the status of the connection to mulesoft
+	health := hc.GetStatus(anypoint.HealthCheckEndpoint)
```

Comment:
The health check is using the same check performed by the client to see if the connection to mulesoft is alive. Should we also check if the connection to condor is good too?

I'm also a little unsure if this is the right way to implement this check. The mulesoft client registers a health check for this endpoint, `anypoint.HealthCheckEndpoint`. And this endpoint just gets the result from that. Is that fine @jcollins-axway  @vivekschauhan

Reply:
This is fine, as the sdk should regularly be validating that the connection is up.  The job is just piggy backing on what the health check library is doing.